src/Silk.Core.Discord/Main.cs

```diff
@@ -0,0 +1,124 @@
+using Silk.Core.Discord.Utilities.HelpFormatter;
+using Silk.Extensions;
+
+namespace Silk.Core.Discord
+{
+    public class Main : IHostedService
+    {
+        public static BotState State { get; private set; } = BotState.Starting;
+        //public static DiscordSlashClient SlashClient { get; } // Soon™ //
+        public static DiscordShardedClient ShardClient { get; private set; }
+        public static string DefaultCommandPrefix { get; } = "s!";
+
+        private static ILogger<Main> _logger;
+        private readonly IServiceProvider _provider;
+
+        public Main(DiscordShardedClient shardClient, ILogger<Main> logger, IServiceProvider provider)
+        {
+            _logger = logger;
+            _provider = provider;
+            ShardClient = shardClient;
+        }
+
+        [MethodImpl(MethodImplOptions.Synchronized)]
+        public static void ChangeState(BotState state)
+        {
+            if (State != state)
+            {
+                _logger.LogDebug("State changed from {State} to {NewState}!", State, state);
+                State = state;
+            }
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Starting service");
+            await InitializeClientAsync();
+            await InitializeCommandsNextAsync();
+            _logger.LogDebug("Connecting to Discord gateway");
+            await ShardClient.StartAsync();
+            _logger.LogInformation("Connected to Discord gateway as {Username}#{Discriminator}", ShardClient.CurrentUser.Username, ShardClient.CurrentUser.Discriminator);
+        }
+        public async Task StopAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Stopping service");
+            _logger.LogDebug("Disconnecting from Discord gateway");
+            await ShardClient.StopAsync();
+            _logger.LogInformation("Disconnected from Discord gateway");
+        }
+
+        private async Task InitializeClientAsync()
+        {
+            _logger.LogDebug("Initializing client");
+
+            await ShardClient.UseCommandsNextAsync(DiscordConfigurations.CommandsNext);
+            await ShardClient.UseInteractivityAsync(DiscordConfigurations.Interactivity);
+
+            SubscribeToEvents();
+        }
+        private void SubscribeToEvents()
+        {
+            //Client.MessageCreated += _services.Get<AutoModInviteHandler>().MessageAddInvites; // I'll fix AutoMod eventually™ ~Velvet, May 3rd, 2021. //
+            var services = _provider;
+            var mediator = services.Get<IMediator>()!;
+
+            // Direct Dispatch //
+            ShardClient.MessageDeleted += services.Get<MessageRemovedHandler>()!.MessageRemoved;
+            ShardClient.GuildMemberAdded += services.Get<MemberAddedHandler>()!.OnMemberAdded;
+            ShardClient.GuildMemberUpdated += services.Get<RoleAddedHandler>()!.CheckStaffRole;
```

Comment:
Does this not get the service until it's first used? If not, it would be better to have your handlers piped in through DI and then consume them within your constructor to initialize everything. I mean, it's what a constructor is for.

Reply:
If you care to read, MoMo on Discord did bring up a perfectly valid argument about depdencies when it comes to the service locator.

> ```cs
> new Main(new DiscordShardedClient(...), new MainLogger(), new SomeServiceProvider(...));
> ```

> is 100% correct/valid code (ignoring any ctor's of those sub types) but actually in reality would be wrong because there hidden dependencies that I will never know or be told (or find in refactorings) because the Service Locator (or provider) hides them all.
In general, the service locator should never be injected into any class ever. It resolves the primary root(s) which then those root(s) stay in memory or in the loop of public static void main until the app terminates. The container could be removed and the app should still function correctly but just using manually created objects instead of the container auto-wiring them all together.
It seems like the handlers are the responsibility of the SharedClient implementation to get injected with and setup. Using mediator to publish messages is fine I would wager (though I would then question what is the purpose of SubscribeToEvents if all it does it work with the client and modify it internally? Should this really be part of the client?) but the service locator usage here shows we're very knowledgeable and interested in the inner workings of the client.

> Also, are we sure SubscribeToEvents is correct, logically? Is it going to keep adding event handlers onto each of those events (I don't recall if += checks to see if the delegate already exists before adding it (or would there be bugs because the service/provided instance of the delegate is technically a different object??))? For example, if we do Start, Stop, Start ... I would expect there to be duplicate delegates on each event (especially the mediator ones because those lambdas are new'ed up on the spot and are different objects [lambdas are just hiding Action/Func objects]). At least that is my first impression.